test/providers/provider_level_zero.cpp

```diff
@@ -0,0 +1,210 @@
+
+        // close and execute the command list
+        ze_result = zeCommandListClose(hCommandList);
+        ASSERT_EQ(ze_result, ZE_RESULT_SUCCESS);
+        ze_result = zeCommandQueueExecuteCommandLists(hCommandQueue, 1,
+                                                      &hCommandList, nullptr);
+        ASSERT_EQ(ze_result, ZE_RESULT_SUCCESS);
+
+        // sync
+        ze_result = zeCommandQueueSynchronize(
+            hCommandQueue, std::numeric_limits<uint64_t>::max());
+        ASSERT_EQ(ze_result, ZE_RESULT_SUCCESS);
+    }
+
+    void CopyDeviceToHostMemory(void *host_ptr, void *device_ptr, size_t size) {
+        ASSERT_NE(host_ptr, nullptr);
+        ASSERT_NE(device_ptr, nullptr);
+
+        ze_command_queue_desc_t commandQueueDesc = {
+            ZE_STRUCTURE_TYPE_COMMAND_QUEUE_DESC};
+        ze_command_list_desc_t commandListDesc = {
+            ZE_STRUCTURE_TYPE_COMMAND_LIST_DESC};
+
+        ze_command_queue_handle_t hCommandQueue;
+        ze_result_t ze_result = zeCommandQueueCreate(
+            hContext, hDevice, &commandQueueDesc, &hCommandQueue);
+        ASSERT_EQ(ze_result, ZE_RESULT_SUCCESS);
+
+        ze_command_list_handle_t hCommandList;
+        ze_result = zeCommandListCreate(hContext, hDevice, &commandListDesc,
+                                        &hCommandList);
+        ASSERT_EQ(ze_result, ZE_RESULT_SUCCESS);
+
+        // copy from device memory to host memory
+        zeCommandListAppendMemoryCopy(hCommandList, host_ptr, device_ptr, size,
+                                      nullptr, 0, nullptr);
+
+        // close and execute the command list
+        zeCommandListClose(hCommandList);
+        zeCommandQueueExecuteCommandLists(hCommandQueue, 1, &hCommandList,
+                                          nullptr);
+
+        zeCommandQueueSynchronize(hCommandQueue,
+                                  std::numeric_limits<uint64_t>::max());
+    }
+
+    ze_driver_handle_t hDriver;
+    ze_device_handle_t hDevice;
+    ze_context_handle_t hContext;
+};
+
+TEST_F(umfLevelZeroProviderTest, basic) {
+    const size_t size = 1024 * 8;
+    const uint32_t pattern = 0xAB;
+
+    // setup params
+    level_zero_memory_provider_params_t params;
+    params.context = hContext;
+    params.device = hDevice;
+    params.memory_type = ZE_MEMORY_TYPE_DEVICE;
+    ze_device_mem_alloc_desc_t alloc_desc = {
+        ZE_STRUCTURE_TYPE_DEVICE_MEM_ALLOC_DESC, nullptr,
+        0, // flags
+        0  // ordinal
+    };
+
+    // create Level Zero provider
+    params.device_mem_alloc_desc = alloc_desc;
+    umf_memory_provider_handle_t provider = nullptr;
+    umf_result_t umf_result = umfMemoryProviderCreate(
+        umfLevelZeroMemoryProviderOps(), &params, &provider);
+    ASSERT_EQ(umf_result, UMF_RESULT_SUCCESS);
+    ASSERT_NE(provider, nullptr);
+
+    const char *message;
+    int32_t error = 0;
+    umfMemoryProviderGetLastNativeError(provider, &message, &error);
+    ASSERT_EQ(error, 0);
```

Comment:
I don't think you should test for that. The return value of umfMemoryProviderGetLastNativeError is unspecified if the previous call was successful.

Reply:
removed